packages/flutter/lib/src/rendering/_first_frame_web.dart

```diff
@@ -0,0 +1,11 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:html' as html;
+
+/// Dispatch an event to signal to the service worker binding that registration
+/// is safe.
+void dispatchFirstFrame(Duration timeStamp) {
+  html.window.dispatchEvent(html.Event('flutter-first-frame'));
```

Comment:
It's a little bit odd to see a call to `html` outside of web_ui (but I like how direct this is). 

How about making the first frame rendered some sort of PlatformMessage that gets intercepted by the `web_ui` engine window [here](https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/window.dart#L562), and broadcast as an html.window event from there?

_(I'm not 100% sure of what needs to be done (or not done) in the native side to noop to an unknown platform message, though)_

Reply:
Disclaimer: I don't know enough about the binding, if @goderbauer is cool with the PR, he's the boss)

Reply:
We could conditionally call the platform channel on the web, otherwise we would need to add a nop impl on every platform I think.

Reply:
I've seen the [Result.notImplemented](https://github.com/flutter/engine/blob/33015c6c7a7d215fba4f3caa9d2888de9658034a/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L189) thing as a default, but I'm not sure how that's handled by the framework later (if it errors out, or just silently noops)

Reply:
Opened https://github.com/flutter/engine/pull/21284

Reply:
I pinged the chat for them to take a look!, going to mention this PR as well, thanks!!

Reply:
_(Resolving this, thanks for considering this comment, Jonah!)_